design/batch_retrieval_design.md

```diff
@@ -1,20 +1,183 @@
+
+3. For each mapped Kubernetes secret, 
+
+   a. The `cyberark-secrets-provider-for-k8s` performs a Batch Retrieval request and retrieves Conjur/DAP secrets.
+
+   b. It updates the Kubernetes secret with the Conjur/DAP secret value.
+
+4. The `cyberark-secrets-provider-for-k8s` Job is completed.
+
+The new Batch Retrieval API will effect #2a. 
+
+### URI
+
+##### POST /secrets/batch
+
+This method retuns a list of secrets, their status code, and their value if an error did not occur. See difference between proposed API and existing API see [Proposed API vs Old API](#proposed-api-vs-old-api) below.
+
+```
+POST /secrets/batch
+
+{
+  "variables": [
+    {"name": "var/path/conjurSecret"},
+    {"name": "var/path/anotherConjurSecret"},
+    ...
+  ]
+}
+```
+
+### Request Body
+
+The request body is a list of all variable IDs (paths to secrets in Conjur/DAP) that need to be retrieved. For example: 
+
+```
+{
+  "variables": [
+    {"name": "var/path/conjurSecret"},
+    {"name": "var/path/anotherConjurSecret"},
+    ...
+  ]
+}
+```
+
+#### Headers
+
+```
+Content-Type: application/json
+Authorization: Token token="eyJkYX...Rhb="
+```
+
+### Response
+
+#### Batch Response
+
+If no Conjur/DAP secrets were able to be retrieved (no 200 code), the response will still return a 200 OK because the [request body is parsable]([https://docs.microsoft.com/en-us/graph/json-batching#:~:text=The%20status%20code%20on%20a,requests%20inside%20the%20batch%20succeeded](https://docs.microsoft.com/en-us/graph/json-batching#:~:text=The status code on a,requests inside the batch succeeded)) and request was handled.
+
+If an invalid request body was supplied, a 400 Bad Request will be returned
+
+| Code            | Description                                                  |
+| --------------- | ------------------------------------------------------------ |
+| 200 OK          | All secret values were retrieved successfully                |
+| 400 Bad Request | Server cannot process request due to client error (invalid input parameters or logical validations) |
+
+#### Individual secret responses
+
+All secret values will be treated as individual entities. If a single secret fails to be retrieved, this will not impact the integrity of the request. Each secret has it's own response code.
+
+| Code             | Description                                                  |
+| ---------------- | ------------------------------------------------------------ |
+| 200 OK           | All secret values were retrieved successfully                |
+| 400 Bad Request  | Server cannot process request due to client error (invalid input parameters or logical validations) |
+| 401 Unauthorized | The request lacks valid authentication credentials           |
+| 403 Forbidden    | The authenticated user lacks the necessary privilege         |
+| 404 Not Found    | Variable does not exist or variable does not have secret value |
```

Comment:
I don't think these messages are in line with standard JSON responses. I think more appropriate would be something like:

Worked OK:
```json
{
  "id": "foo/bar",
  "success": true,
  "value": "somevalue"
}
```

And for errors:
```json
{
  "id": "foo/bar",
  "success": false,
  "error": {
    "code": 404,
    "message": "Variable not found"
  }
}
```

Google uses this pattern quite a bit and you can find more info about it here: https://www.jsonrpc.org/specification#error_object

Reply:
Do you think there is added value of having an extra field for success named `"code": 200` instead of or alongside `"success": true`?

Reply:
So it would be:
```
  "variables": [
  	{
      "id": "var/path/name1",
      "code": 200,
      "value": "superSecret"
    },
    {
      "id": "var/path/name2",
      "error": {
        "code": 403,
        "message": "Forbidden"
      }
    },
...
```
where in-place of the `"success": true` we have error or no error.

Reply:
You have the code in the base in case of success, but in the error object in case of failure.
This is inconsistent.

Reply:
do we not see value in returning the code if success in general?